feat: issue_15_tailwindcss

yoshikawa-akr opened the pull request:
# [issue No.15](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#15)

**issue**
モーダル内に formをおいて、input(type number)値、input(type text)、を新たに作ったsubmitボタン押下で リクエスト、 APIから値を取ってきてください

## [CodeSandBox](https://githubbox.com/yoshikawa-akr/js-lessons/tree/develop/issue_15_tailwindcss) / Latest commit: https://github.com/yoshikawa-akr/js-lessons/pull/16/commits/e0507b1fbbc62cb2548e7e6c0a71f8c3629715bf

## Implementation concerns
- Whether there are any behavior problems due to rewriting issue 15 with TailwindCSS.

yoshikawa-akr committed:
feat: Add tailwindcss own class

yoshikawa-akr committed:
feat: Adding and removing classes

yoshikawa-akr committed:
feat: Change transition setting

yoshikawa-akr committed:
add: Add gif and png files

yoshikawa-akr committed:
feat: Add class for centering to loadingImg

haru-programming commented on issue_15_tailwindcss/main.js:
```diff
@@ -17,7 +18,7 @@ function removeLoading() {
 }
 
 function removeShowClassOfmodal() {
-    modal.classList.remove('show');
+    modal.classList.add('opacity-0', 'invisible');
 }
```
[ IMO ]
With the change to tailwind CSS, I thought the function names did not match the content.

yoshikawa-akr replied:
https://github.com/yoshikawa-akr/js-lessons/pull/16/commits/909740e0b049560484a5ebe62b067ee7dc9ce9c9
Thank you for pointing this out.
You are exactly right:sparkles:
This time, TailwindCSS has put together the styles as custom CSS, so I have changed the function names and their processing accordingly.

haru-programming reviewed (commented):
[ Q. ]
`document.getElementById('js-form').reset();`

This is the first time I have heard of `reset()`. Thank you very much.

Do I need to reset the value here?
It would be a great learning experience  if you could tell me 🙇‍♀️

haru-programming commented:
[ comment ]
I thought the operation was fine!

Saki-Ota commented on issue_15_tailwindcss/index.html:
```diff
@@ -1,35 +1,38 @@
+        <form method="post" id="js-form">
+            <div class="modal__element flex justify-center items-center mb-8 text-black">
+                <label for="js-input-number" class="modal__input-label pr-6 cursor-pointer">数字</label>
+                <input type="number" name="number" class="modal__input text-base border border-black bg-transparent p-1" id="js-input-number">
```
FYI
As the class content is repeating, you may want to use @apply method to re use the class. 
Here is the reference 
[tailwind @apply](https://tailwindcss.com/docs/reusing-styles#extracting-classes-with-apply)
[Functions & Directives](https://tailwindcss.com/docs/functions-and-directives#layer)

```suggestion
// in CSS file (if you create one)
@layer components {
.modal_input {
  @apply  text-base border border-black bg-transparent p-1
 }
}
```
I am still learning how to use @apply and @layer method so that I wouldn't say this could be the best way to do, but I thought it would be something you might want to have a look.

yoshikawa-akr replied:
https://github.com/yoshikawa-akr/js-lessons/pull/16/commits/afe8c1d80cd0bda85b98de40865f4c3b481075cb
Thank you for your comment.
Such a useful feature! I didn't know about it.
I put together some classes.

Saki-Ota replied:
I checked how you refactored your tailwind CSS and it looks pretty good. I learned from you too. Thanks

nakagawamai commented:
MUST
I think class names that are no longer used should be removed as TailwindCSS is supported.
Because there is no corresponding stylesheet.

Exsample : `modal__element` ,`modal__input`, `modal__input-label` and more.

nakagawamai commented:
LGTM
I thought the hover on the submit button worked great!
https://user-images.githubusercontent.com/99260576/209527917-0c6d8ffd-596d-48b6-8553-daf6c176f940.mov

yoshikawa-akr committed:
remove: Remove description to reset form values

Because the process was unnecessary for this implementation.

yoshikawa-akr committed:
remove: Remove unused class and id

yoshikawa-akr committed:
feat: Integration of common styles

yoshikawa-akr committed:
feat: Support for custom CSS in JS

yoshikawa-akr commented:
> [ Q. ] `document.getElementById('js-form').reset();`
> 
> This is the first time I have heard of `reset()`. Thank you very much.
> 
> Do I need to reset the value here? It would be a great learning experience if you could tell me 🙇‍♀️


ce55ce58871be8e2820588f9db283ff355d66e41
I had written this with the intention that when the modal is displayed multiple times after form submission, the previous value is reset.
However, in this case, the modal is never displayed more than once, so I thought it would be unnecessary.
Thanks to your question, it gave me a chance to review it.
Thank you!

yoshikawa-akr commented:
> MUST I think class names that are no longer used should be removed as TailwindCSS is supported. Because there is no corresponding stylesheet.
> 
> Exsample : `modal__element` ,`modal__input`, `modal__input-label` and more.

https://github.com/yoshikawa-akr/js-lessons/pull/16/commits/3bf6c0f788fe2ead7b29a6cec34e0dae1919eba2
Thank you for pointing this out.
You are right, I should remove the unused classes.
Fixed.

nakagawamai commented:
Q.
Why is the following not TailwindCSS? If you have a reason, let me know!
```CSS
:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;
  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;
  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}
a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}
body {
  margin: 0;
  display: flex;
  place-items: center;
  min-width: 320px;
  min-height: 100vh;
}
h1 {
  font-size: 3.2em;
  line-height: 1.1;
}
#app {
  max-width: 1280px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
}
.logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
}
.logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}
.logo.vanilla:hover {
  filter: drop-shadow(0 0 2em #f7df1eaa);
}
.card {
  padding: 2em;
}
.read-the-docs {
  color: #888;
}
button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}
@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}
```

yoshikawa-akr committed:
remove: Remove excess styles

Removed styles generated when vite was installed.

yoshikawa-akr committed:
refactor: Add padding to buttons

yoshikawa-akr commented:
> Q. Why is the following not TailwindCSS? If you have a reason, let me know!
> 
> ```css
> :root {
>   font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
>   font-size: 16px;
>   line-height: 24px;
>   font-weight: 400;
>   color-scheme: light dark;
>   color: rgba(255, 255, 255, 0.87);
>   background-color: #242424;
>   font-synthesis: none;
>   text-rendering: optimizeLegibility;
>   -webkit-font-smoothing: antialiased;
>   -moz-osx-font-smoothing: grayscale;
>   -webkit-text-size-adjust: 100%;
> }
> a {
>   font-weight: 500;
>   color: #646cff;
>   text-decoration: inherit;
> }
> a:hover {
>   color: #535bf2;
> }
> body {
>   margin: 0;
>   display: flex;
>   place-items: center;
>   min-width: 320px;
>   min-height: 100vh;
> }
> h1 {
>   font-size: 3.2em;
>   line-height: 1.1;
> }
> #app {
>   max-width: 1280px;
>   margin: 0 auto;
>   padding: 2rem;
>   text-align: center;
> }
> .logo {
>   height: 6em;
>   padding: 1.5em;
>   will-change: filter;
> }
> .logo:hover {
>   filter: drop-shadow(0 0 2em #646cffaa);
> }
> .logo.vanilla:hover {
>   filter: drop-shadow(0 0 2em #f7df1eaa);
> }
> .card {
>   padding: 2em;
> }
> .read-the-docs {
>   color: #888;
> }
> button {
>   border-radius: 8px;
>   border: 1px solid transparent;
>   padding: 0.6em 1.2em;
>   font-size: 1em;
>   font-weight: 500;
>   font-family: inherit;
>   background-color: #1a1a1a;
>   cursor: pointer;
>   transition: border-color 0.25s;
> }
> button:hover {
>   border-color: #646cff;
> }
> button:focus,
> button:focus-visible {
>   outline: 4px auto -webkit-focus-ring-color;
> }
> @media (prefers-color-scheme: light) {
>   :root {
>     color: #213547;
>     background-color: #ffffff;
>   }
>   a:hover {
>     color: #747bff;
>   }
>   button {
>     background-color: #f9f9f9;
>   }
> }
> ```

https://github.com/yoshikawa-akr/js-lessons/pull/16/commits/2ca12019aa77f211ddd512a6d024dab3c2331edc
https://github.com/yoshikawa-akr/js-lessons/pull/16/commits/e0507b1fbbc62cb2548e7e6c0a71f8c3629715bf
Thank you for pointing this out.
I think this is probably a style that was generated when vite was installed and was left behind.
I thought it was the initial description of TailwindCSS, so I left it as is.
Since it was unnecessary, I deleted it and added a new style :raised_hands:
